### PR TITLE
Change getComposedRanges to have options as input

### LINKED ISSUE #176: Change `getComposedRanges`'s `shadowRoots` parameter from rest parameter to a dictionary
Current [spec](https://w3c.github.io/selection-api/#ref-for-dom-selection-getcomposedranges-1) of `getComposedRanges` API has single `shadowRoots` parameter which is a rest parameter. We should change it to a dictionary that contains an array of shadow roots. Similar to https://html.spec.whatwg.org/#gethtmloptions. It'd be good to be consistent.

Proposed change:

```
sequence<StaticRange> getComposedRanges(optional GetComposedRangesOptions options = {});

dictionary GetComposedRangesOptions {
  sequence<ShadowRoot> shadowRoots = [];
};
```

### PR DESCRIPTION
**Pull request by dizhang168**
Closes #176

Implementation commitment:

 * [x] WebKit (https://github.com/w3c/selection-api/issues/176#issuecomment-2362070387) (caveat web compat)
 * [x] Chromium (https://issues.chromium.org/issues/355577223, https://github.com/w3c/selection-api/issues/176#issuecomment-2118432134)
 * [x] Gecko (https://github.com/w3c/selection-api/issues/176#issuecomment-2256467136)

***
<a href="https://pr-preview.s3.amazonaws.com/dizhang168/selection-api/pull/331.html" title="Last updated on Sep 30, 2024, 3:19 PM UTC (4efbda8)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/selection-api/331/45f8fe8...dizhang168:4efbda8.html" title="Last updated on Sep 30, 2024, 3:19 PM UTC (4efbda8)">Diff</a>

### CODE REVIEW

**Comment by dizhang168** [comment]
@rniwa @sanketj

Please help review, thanks!

**Review by sanketj — APPROVED** [review]
This looks reasonable to me from a Chromium perspective and in terms of spec consistency with other similarly shaped APIs. The web platform design principles also support this updated syntax: https://www.w3.org/TR/design-principles/#prefer-dictionaries.

However, it might be good to get an official resolution on this change from the Editing WG before landing the PR. cc: @johanneswilm

**Comment by dizhang168** [comment]
The WPT test changes are at:
https://github.com/web-platform-tests/wpt/pull/47523

**Review by dandclark — COMMENTED** [review]
This should also change the definition of this function to refer to `options.shadowRoots` instead of `shadowRoots`.

**Review by dandclark — APPROVED** [review]

File: index.html [hunk 1]
```diff
@@ -405,7 +409,7 @@ <h2>
             <li>Otherwise, let <var>startNode</var> be [=range/start node=] of the [=range=] associated with [=this=],
               and let <var>startOffset</var> be [=range/start offset=] of the [=range=].</li>
             <li>While <var>startNode</var> is a [=node=], <var>startNode</var>'s [=tree/root=] is a [=shadow root=],
-              and <var>startNode</var>'s [=tree/root=] is not a [=shadow-including inclusive ancestor=] of any of <var>shadowRoots</var>,
+              and <var>startNode</var>'s [=tree/root=] is not a [=shadow-including inclusive ancestor=] of any of <var>options.shadowRoots</var>,
```
Comment (annevk) [inline index.html]: This (and below) needs different syntax. You want it look something like this:

> _options_["`shadowRoots`"]

Ideally with `shadowRoots` linking back to its IDL definition.

Comment (dizhang168) [reply index.html]: Thanks for the review.  I can change to format _options_[_shadowRoots_] by using `<var>options</var>[<var>shadowRoots</var>]`.

However, I am not sure how to link back to the IDL definition, I tried:
`<var>options</var>[<var data-x="dom-getcomposedrangesoptions-shadowroots">shadowRoots</var>]` 
and
`sequence&lt;ShadowRoot&gt; <dfn data-x="dom-getcomposedrangesoptions-shadowroots">shadowRoots</dfn> = [];`, but that didn't work.

Comment (annevk) [reply index.html]: shadowRoots here is not a variable, it's a string. I think something like

`<var>options</var>["{{GetComposedRangeOptions/shadowRoots}}"]`

should work assuming this is Bikeshed.

Comment (annevk) [reply index.html]: You need to remove the whitespace between the variable and the brackets to get the correct formatting.

Comment (dizhang168) [reply index.html]: Thanks for the patience and review. I didn't realize the previous change added the space, updated.

**Review by annevk — APPROVED** [review]

**Pull request merged**